enhancements/ingress/request-timeout-configuration.md

```diff
@@ -0,0 +1,183 @@
+---
+title: request-timeout-configuration
+authors:
+  - "@rfredette"
+reviewers:
+  - "@miciah"
+  - "@danehans"
+  - "@frobware"
+  - "@sgreene570"
+  - "@knobunc"
+  - "@miheer"
+  - "@candita"
+  - "@alebedev87"
+approvers:
+  - "@miciah"
+  - "@frobware"
+  - "@knobunc"
+creation-date: 2021-06-17
+last-updated: 2021-06-28
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced|informational
+see-also:
+replaces:
+superseded-by:
+---
+
+# Request Timeout Configuration
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This document proposes the addition of several API fields into the ingress
+controller API, allowing the cluster admin to adjust how long routers will hold
+open connections that are awaiting a response.
+
+## Motivation
+
+Customers need the ability to adjust how long HAProxy holds connections open,
+either by extending the timeout to accomodate slower backends or clients, or by
+shortening the timeout, allowing connections to be closed more aggressively.
+
+### Goals
+
+Allow admins to configure various HAProxy connection timeout parameters
+
+### Non-Goals
+
+TODO
+
+## Proposal
+
+Add the following configurable variables to the ingresscontroller API:
+```yaml
+spec:
+  tuningOptions:
+    clientTimeout: "30s"
```

Comment:
Is there a better way to express these timeouts than strings? If not, we should make sure to define the format and what limits HAProxy might place on these values come API design time.

Reply:
(my main concern is someone specifying a value that breaks HAProxy reloads such as "99999d" on accident)

Reply:
I think it's tough to validate the field at the API level because I don't know of a good way to allow users to specify units while also validating that the timeout is not longer than the max allowed value.

If validation at the API level is preferred, input could be taken as an integer, assuming a particular time unit (my inclination would be `ms`), which would make it easy to have min/max values, but without being able to specify the units, there's more chance the user making a conversion error when inputting the timeouts.

Another option is to follow the same policy as the hard-stop-after timeout, where the ingress operator accepts any valid time as a string in the HAProxy time format, but if the value is higher than the max allowed, it gets truncated to the max value (i.e. `99999d` would become `2147483647ms`) and an error is logged.

Between the two, I'd prefer the latter behavior since it improves readability and reduces the chance of user error, but I'm open to other perspectives

Reply:
Ah, right, I forgot about our timeout value clipping logic. Maybe we could take this oppurtunity to move that to a more formal location (I think in the past we talked about moving the HAProxy timeout clip function to library-go, or something along those lines).

Reply:
aha! https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/deployment.go#L1262-L1263

Reply:
By the way, https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-conventions is relevant here.  It looks like the names you chose are consistent with the guidance there.  (There is some interesting nuance between "timeouts" and "deadlines",  but I think these are all timeouts.)

Reply:
Good to know; I was not aware of the deadline vs timeout guidance, but these fields do seem to align with the timeout definition.